### PR TITLE
Exponential back-off support for Google PaLM api

### PR DESCRIPTION
**Pull request by mtronci**
This PR adds exponential back-off to the Google PaLM api to gracefully handle rate limiting errors.

### CODE REVIEW

File: langchain/embeddings/google_palm.py [hunk 1]
```diff
@@ -6,6 +6,38 @@
...
+    def deco_retry(f):
+        @wraps(f)
+        def f_retry(*args, **kwargs):
+            mtries, mdelay = tries, delay
+            while mtries > 1:
+                try:
+                    return f(*args, **kwargs)
+                except ExceptionToCheck:
+                    msg = "%s, Retrying in %d seconds..." % (str(ExceptionToCheck), mdelay)
+                    print(msg)
```
Comment (vowelparrot) [inline langchain/embeddings/google_palm.py:33]: Should use `logger.info()` instead

**Review by vowelparrot — COMMENTED** [review]
Would be preferable to reuse / share / extend the `completion_with_retry` decorators that use `tenacity`. You can reference these at `langchain/{chat_models|embeddings|llms}/openai.py` and in the confluence dataloader

File: langchain/embeddings/google_palm.py (hunk 1 above)
Comment (mtronci) [reply langchain/embeddings/google_palm.py:33]: many thanks for reviewing.
I'll look to replicate a similar approach.

**Comment by mtronci** [comment]
I have reworked the changes, and also added retry support to the PaLM chat.
I have gone for a simple approach, it is not exactly the same as the OpenAI one.
The decorator is stored in a class attribute, and then used to decorate the relevant methods.

**Review by vowelparrot — APPROVED** [review]

**Comment by brotchie** [comment]
Thanks so much for adding this @mtronci and to all reviewers 🙏

**Comment by mtronci** [comment]
thanks all for reviewing and merging!

**Pull request merged**